ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -1298,6 +1324,93 @@ private void validateAndClose() {
+    info("========================= Summary =========================");
+    for (Map.Entry<String, Set<String>> entry : dcToSealedPartitions.entrySet()) {
+      info("Dc {} has {} sealed partitions.", entry.getKey(), entry.getValue().size());
+      sealedPartitionsInCluster.addAll(entry.getValue());
+    }
+    info("========================= Sealed Partitions across All DCs =========================");
+    info("Total number of sealed partitions in cluster = {}", sealedPartitionsInCluster.size());
+    info("Sealed partitions are {}", sealedPartitionsInCluster.toString());
+    if (!nodeToNonExistentReplicas.isEmpty()) {
+      info("Following {} nodes have sealed replica that are not actually present", nodeToNonExistentReplicas.size());
+      for (Map.Entry<String, Set<String>> entry : nodeToNonExistentReplicas.entrySet()) {
+        info("{} has non-existent replicas: {}", entry.getKey(), entry.getValue().toString());
+      }
+    }
+    info("Successfully aggregate sealed from cluster {} in Helix", clusterName);
+    return sealedPartitionsInCluster;
+  }
+
+  /**
+   * Get sealed partitions from given datacenter.
+   * @param dc the datacenter where sealed partitions come from.
+   * @param dcToSealedPartitions a map to track sealed partitions in each dc. This entry associated with given dc will
+   *                             be populated in this method.
+   * @param nodeToNonExistentReplicas a map to track if any replica is in sealed list but not actually on local node.
+   */
+  private void getSealedPartitionsInDc(Datacenter dc, Map<String, Set<String>> dcToSealedPartitions,
+      Map<String, Set<String>> nodeToNonExistentReplicas) {
+    String dcName = dc.getName();
+    dcToSealedPartitions.put(dcName, new HashSet<>());
+    HelixAdmin admin = adminForDc.get(dcName);
+    Set<String> allInstancesInHelix = new HashSet<>(admin.getInstancesInCluster(clusterName));
+    for (DataNodeId dataNodeId : dc.getDataNodes()) {
+      DataNode dataNode = (DataNode) dataNodeId;
+      Set<String> replicasOnNode = staticClusterMap.getReplicas(dataNode)
+          .stream()
+          .map(replicaId -> replicaId.getPartitionId().toPathString())
+          .collect(Collectors.toSet());
+      String instanceName = getInstanceName(dataNode);
+      ensureOrThrow(allInstancesInHelix.contains(instanceName), "Instance not present in Helix " + instanceName);
+      InstanceConfig instanceConfig = admin.getInstanceConfig(clusterName, instanceName);
+      List<String> sealedReplicas = instanceConfig.getRecord().getListField(ClusterMapUtils.SEALED_STR);
+      if (sealedReplicas != null) {
+        for (String sealedReplica : sealedReplicas) {
+          info("Replica {} is sealed on {}", sealedReplica, instanceName);
+          dcToSealedPartitions.get(dcName).add(sealedReplica);
+          if (!replicasOnNode.contains(sealedReplica)) {
+            logger.warn("Replica {} is in sealed list but not on node {}", sealedReplica, replicasOnNode);
```

Comment:
replicasOnNode is a set of all replicas on a node, you may want to log the node id here.

Reply:
Good catch, fixed